plugins/kubernetes-backend/src/service/KubernetesProxy.ts

```diff
@@ -123,7 +126,25 @@ export class KubernetesProxy {
       }
 
       const middleware = await this.getMiddleware(req);
-      middleware(req, res, next);
+      if (
+        req.header('connection')?.toLowerCase() === 'upgrade' &&
+        req.header('upgrade')?.toLowerCase() === 'websocket'
+      ) {
+        type IPartialUpgradeHandler = (
+          ...args: Partial<
+            Parameters<
+              Server<IncomingMessage, ServerResponse<IncomingMessage>>['ws']
+            >
+          >
+        ) => void;
```

Comment:
Oof.

Seems to me that you could instead change `getMiddleware` to return a `http-proxy-middleware.RequestHandler` instead of an `express.RequestHandler`. Oh and of course also changing `middlewareForClusterName` accordingly too. Then you already have the `upgrade` method on the type without all of this trickery.

Only thing is that it's actually optional, but you're casting that away anyway, so you'll be able to call it as

```ts
middleware.upgrade!(req, res.socket, undefined);
```

removing a ton of code

Reply:
May also be nice with a comment explaining why the if statement exists at all.